packages/block-editor/src/components/block-list/block.js

```diff
@@ -307,6 +311,26 @@ const applyWithDispatch = withDispatch( ( dispatch, ownProps, registry ) => {
 				canInsertBlockType,
 			} = registry.select( blockEditorStore );
 
+			/**
+			 * A block is "empty" if all of its content attributes are unmodified.
+			 *
+			 * @param {WPBlock} block The block to check.
+			 * @return {boolean} Whether the block is empty.
+			 */
+			function isBlockEmpty( block ) {
```

Comment:
How is this different to the existing `isUnmodifiedBlock` selector that we have?

Reply:
It's also discussed here: https://github.com/WordPress/gutenberg/pull/65262#discussion_r1756466550.

An "empty" block could still have some attributes modified. For instance, a heading block could have its content empty but also aligned to the center. The intent is to assume that the user might want to remove/reset the block when it's empty instead of moving it up to the parent when it has contents. Note that this is not the only solution but a proposed trade-off to fix https://github.com/WordPress/gutenberg/issues/65174.

Reply:
So an "empty" block is a block whose "content" role attribute are unmodified. 

I think this function should be removed `isBlockContentUnmodified` and probably be moved to `blocks` package as a private selector maybe, that way it's placed closed to the other `isUnmodifiedBlock`.

Now for the behavior itself, I understand the tradeoff, I think it's an ok tradeoff if this flow only applies to "text" blocks which is probably true. But I'm concerned about the impact on third-party "text" blocks because we just stabilized "content" role and I'm sure no one is actually using it. To make sure I understand the tradeoff better, what happens if a third-party text block is not assigning `role="content"` properly to its attribute?

Reply:
> I think this function should be removed `isBlockContentUnmodified` and probably be moved to `blocks` package as a private selector maybe, that way it's placed closed to the other `isUnmodifiedBlock`.

This PR also adds a private `isAttributeUnmodified` function in the `blocks` package before the content role is stabilized. Since that it's now stabilized, I think we can move `isBlockEmpty` there too.

> what happens if a third-party text block is not assigning `role="content"` properly to its attribute?

If there's no `content` role attributes, then it's also considered "empty". Then it'll follow the same steps as mentioned above.

Reply:
>If there's no content role attributes, then it's also considered "empty". Then it'll follow the same steps as mentioned above.

So content loss? I guess there's always "undo" but I wonder if we can find a better way here.

Reply:
I'm wondering if something like this would work better:

 - If a block has "content" attribute, use `isBlockContentUnmodified`, if not fallback to `isBlockUnmodified`

Reply:
I think that's a fair trade-off for now. I just rebased and updated the branch!